distributed/dashboard/components/scheduler.py

```diff
@@ -438,6 +439,116 @@ def name(address):
+
+    def __init__(self, scheduler, **kwargs):
+        with log_errors():
+            self.last = 0
+            self.scheduler = scheduler
+
+            es = [p for p in self.scheduler.plugins if isinstance(p, TaskStreamPlugin)]
+            if not es:
+                self.plugin = TaskStreamPlugin(self.scheduler)
+            else:
+                self.plugin = es[0]
+
+            self.actions = [
+                "transfer",
+                "compute",
+            ]  # "disk_write", "disk_read", "deserialize"
+
+            data = {
+                "x": [("nokey", "transfer"), ("rk", "compute")],
+                "counts": [0.2, 0.1],
+            }
+
+            self.source = ColumnDataSource(data=data)
+
+            fig = figure(
+                title="Computer / Transfer Ratio",
+                tools="",
+                id="bk-ratio-by-key-plot",
+                name="ratio_by_key",
+                x_range=FactorRange(),
+                **kwargs,
+            )
+
+            rect = fig.vbar(
+                source=self.source, x="x", top="counts", width=0.9, line_color="black"
+            )
+            fig.yaxis[0].formatter = NumeralTickFormatter(format="0.0s")
+
+            fig.yaxis.ticker = AdaptiveTicker(**TICKS_1024)
+            fig.xaxis.major_label_orientation = -math.pi / 12
+            rect.nonselection_glyph = None
+
+            fig.xaxis.minor_tick_line_alpha = 0
+            fig.ygrid.visible = False
+
+            fig.toolbar.logo = None
+            fig.toolbar_location = None
+
+            # hover = HoverTool()
+            # hover.tooltips = "@name: @transfer objects"
+            # hover.tooltips = """
+            # <div>
+            #     <p><b>Name:</b> @name</p>
+            #     <p><b>Transfer:</b> @transfer objects</p>
+            #     <p><b>Compute:</b> @compute objects </p>
+            #     <p><b>Deserialize:</b> @deserialize objects </p>
+            # </div>
+            # """
+            # hover.point_policy = "follow_mouse"
+            # fig.add_tools(hover)
+
+            fig.y_range.start = 0
+            fig.x_range.range_padding = 0.1
+            fig.xaxis.major_label_orientation = 1
+            fig.xgrid.grid_line_color = None
+
+            self.fig = fig
+
+    @without_property_validation
+    def update(self):
+        with log_errors():
+            # transfer = defaultdict(float)
+            key_vals = defaultdict(float)
+            totals = defaultdict(float)
+
+            for task in self.scheduler.get_task_stream():
+                name = key_split(task["key"])
+                startstops = task["startstops"]
+
+                for data in startstops:
+
+                    act = data["action"]
+
+                    start = data["start"]
+                    stop = data["stop"]
+                    time = stop - start
+
+                    key_vals[(name, act)] += time
+                    totals[act] += time
```

Comment:
This walks over all tasks and recomputes things every 500ms.  This might be ok, but it might be nicer to instead track some of these metrics on the `TaskPrefix` task in the scheduler during normal operation, and then reflect those numbers here as necessary.

In general with dashboard plots it's better if they do as little work assembling data as possible.  The same things that we humans will want to know are also things that the scheduler will want to know to make better scheduling decisions in the future.  If we can do this work in the scheduler itself just as cheaply (or more cheaply in this case) then I recommend that we do so.

It looks like we already track the duration in every TaskGroup.  Here is a snippet of code from distributed/scheduler.py

```python
            if startstops:
                L = [
                    (startstop["start"], startstop["stop"])
                    for startstop in startstops
                    if startstop["action"] == "compute"
                ]
                if L:
                    compute_start, compute_stop = L[0]
                else:  # This is very rare
                    compute_start = compute_stop = None
            else:
                compute_start = compute_stop = None

            #############################
            # Update Timing Information #
            #############################
            if compute_start and ws.processing.get(ts, True):
                # Update average task duration for worker
                old_duration = ts.prefix.duration_average or 0
                new_duration = compute_stop - compute_start
                if not old_duration:
                    avg_duration = new_duration
                else:
                    avg_duration = 0.5 * old_duration + 0.5 * new_duration

                ts.prefix.duration_average = avg_duration
                ts.group.duration += new_duration
```

Maybe we can do something similar there with transfers (and all other components) and manage state that way.